src/tx.rs

```diff
@@ -579,9 +611,19 @@ impl TransactionInner {
 		}
 		// Pop the most recent savepoint from the stack
 		let savepoint = self.savepoint_stack.pop().ok_or(Error::NoSavepoint)?;
-		// Restore the transaction state to the savepoint
-		self.readset = savepoint.readset;
-		self.scanset = savepoint.scanset;
+		// Restore the readset to the savepoint
+		self.readset.pin().clear();
+		// Clone the readset for the savepoint
+		for key in savepoint.readset.pin().iter() {
+			self.readset.pin().insert(key.clone());
+		}
```

Comment:
Each loop iteration calls `pin()` could be:

```suggestion
let target = self.readset.pin();
for key in savepoint.readset.pin().iter() {
    target.insert(key.clone());
}
```